### ISSUE TITLE
Some smaller AES implementations

### ISSUE THREAD

**Issue by akdjka** [kokke/tiny-AES128-C#17]
You asked for smaller AES implementations, here you are:
http://www.ti.com/tool/aes-128
http://dominik.cc/projekty/avr-aes/

Warning: They have the same timing vulnerability in GF multiplication.

**Comment by cmcqueen** [comment]
See also my [aes-min](https://github.com/cmcqueen/aes-min).

**Comment by DamonHD** [comment]
You might like to look at opentrv/OTAESGCM also.

**Comment by kokke** [comment]
Thank you both very much for your links :)

I tried compiling one of these libraries a few weeks back, and it was considerably smaller than this one!
I will read the code ASAP and try to see if I can learn some tricks to slim down this library :)

I'll keep this issue open for further additions. I will see if I can link directly to this issue in the readme text, so it can be used as a way for peopleto contribute knowledge of smaller implementations :)

**Comment by NateZimmer** [comment]
dont' suppose you have a 256 version anywhere?

**Comment by kokke** [comment]
@NateZimmer this project now supports AES192 and AES256 FYI

**Comment by waldyrious** [comment]
> You might like to look at opentrv/OTAESGCM also.

Link, for convenience :) https://github.com/opentrv/OTAESGCM

**Comment by sdrsdr** [comment]
TI version does not support streaming modes and does not support different key sizes .. this trims the code a lot.. we can do it here also, I think, if we're compiling  for 128 bit keys, but I think we should keep the streaming modes as they are basically few function calls in a loop

**Comment by kokke** [comment]
Hi @sdrsdr 

I think the other mentioned libraries may be slightly smaller for CBC-mode and ECB, but not by much. It's been a few years since I compiled and compared binary sizes, but to the best of my knowledge, only [aes-min](https://github.com/cmcqueen/aes-min) is significantly smaller than this library. 

I am not sure what the current status is because it's been so long since I compared last.
[aes-min](https://github.com/cmcqueen/aes-min) doesn't output a single object-file, so it isn't immediately obvious to me how to compare the binary sizes.

As far as I remember [aes-min](https://github.com/cmcqueen/aes-min) has an option to only use a single key. This means the key-.expansion can be pre-computed, hard-coded and placed in ROM. This trick does a big difference. 

For this library, when only using CTR-mode, GCC will automatically remove the code used for decryption, because it is unused. That is why CTR-mode takes up less than 1KB when compiled for the ARM Thumb ISA :)

**Comment by sdrsdr** [comment]
If we can provide the options we probably should. Relying on the compiler/linker to clean up the code is optimistic at best :-)

**Comment by cmcqueen** [comment]
[aes-min](/cmcqueen/aes-min) does not provide code for any of the encryption modes, only the basic encryption/decryption block operation. It does provide for using a pre-computed key schedule (with a function to compute the key schedule if needed) as well as on-the-fly calculation of the key schedule. [aes-min](/cmcqueen/aes-min) is intended to be a "bare bones" AES-128 building block for tiny embedded systems.

**Comment by kokke** [comment]
@sdrsdr I agree with you that it is optimistic to rely on the compiler. I have had the "pleasure" of using some old shitty KEIL compilers that could not remove dead code like this function.

`InvCipher` and friends should be wrapped in
 ```C
#if (defined(CBC) && (CBC == 1)) || (defined(ECB) && (ECB == 1))
...
#endif
```

**Comment by odzhan** [comment]
I've implemented CTR mode: https://github.com/odzhan/tiny_aes

**Comment by monolifed** [comment]
I made key length determine AES variant, and it became larger:
https://github.com/monolifed/tiny-AES-c
